applications/tari_merge_mining_proxy/src/main.rs

```diff
@@ -43,6 +43,146 @@ use std::convert::Infallible;
+#[cfg(feature = "winservice")]
+#[cfg(target_os = "windows")]
+use tokio::runtime::Runtime;
+#[cfg(feature = "winservice")]
+#[cfg(target_os = "windows")]
+use windows_service::{
+    define_windows_service,
+    service::{ServiceControl, ServiceControlAccept, ServiceExitCode, ServiceState, ServiceStatus, ServiceType},
+    service_control_handler::{self, ServiceControlHandlerResult},
+    service_dispatcher,
+    Result as ServiceResult,
+};
+#[cfg(feature = "winservice")]
+#[cfg(target_os = "windows")]
+const SERVICE_NAME: &str = "tari_merge_mining_proxy";
+#[cfg(feature = "winservice")]
+#[cfg(target_os = "windows")]
+const SERVICE_TYPE: ServiceType = ServiceType::OWN_PROCESS;
+#[cfg(feature = "winservice")]
+#[cfg(target_os = "windows")]
+define_windows_service!(ffi_service_main, service_main);
+
+#[cfg(feature = "winservice")]
+#[cfg(target_os = "windows")]
+pub fn service_main(_arguments: Vec<OsString>) {
+    let dir = env::current_exe().unwrap();
+    env::set_current_dir(&dir).unwrap();
+
+    if let Err(_e) = run_service() {
+        // Handle the error
+    }
+}
+
+#[cfg(feature = "winservice")]
+#[cfg(target_os = "windows")]
+async fn shutdown_signal(signal: ShutdownSignal) {
+    let _result = signal.await;
+}
+
+#[cfg(feature = "winservice")]
+#[cfg(target_os = "windows")]
+pub fn run_service() -> ServiceResult<()> {
+    let mut rt = Runtime::new().unwrap();
+    let mut shutdown = Shutdown::new();
+    let signal = shutdown.to_signal();
+
+    let event_handler = move |control_event| -> ServiceControlHandlerResult {
+        match control_event {
+            ServiceControl::Stop => {
+                let _result = shutdown.trigger();
+                ServiceControlHandlerResult::NoError
+            },
+            ServiceControl::Interrogate => ServiceControlHandlerResult::NoError,
+            _ => ServiceControlHandlerResult::NotImplemented,
+        }
+    };
+
+    let status_handle = service_control_handler::register(SERVICE_NAME, event_handler)?;
+
+    let mut next_status = ServiceStatus {
+        service_type: SERVICE_TYPE,
+        current_state: ServiceState::Running,
+        controls_accepted: ServiceControlAccept::STOP,
+        exit_code: ServiceExitCode::Win32(0),
+        checkpoint: 0,
+        wait_hint: Duration::default(),
+        process_id: None,
+    };
+    status_handle.set_service_status(next_status)?;
+
+    let config = initialize().unwrap();
+
+    let addr = config.proxy_host_address;
+    println!("Listening on {}...", addr);
+
+    let config = MergeMiningProxyConfig::from(config);
+    let xmrig_service = MergeMiningProxyService::new(config, BlockTemplateRepository::new());
+    let service = make_service_fn(|_conn| future::ready(Result::<_, Infallible>::Ok(xmrig_service.clone())));
+
+    let _result = rt.block_on(async {
+        let _result = Server::bind(&addr)
+            .serve(service)
+            .with_graceful_shutdown(shutdown_signal(signal))
```

Comment:
nit: I would rather keep this function local instead of using a "dangling / loosey goosey" function
```suggestion
            .with_graceful_shutdown(signal.map(|_|()))
```

Reply:
Thanks, will give that a try.